design/batch_retrieval_design.md

```diff
@@ -2,19 +2,140 @@
+| 3    | Stay as is                                                   | No additional work needed                                    | Bad UX                                             | -                     |
 
 (*) Fallback solution: use this solution as a safety net, only if the original Batch Retrieval request fails
 
 (**) This solution can be broken up in two: 1. Create a new API endpoint or 2. use existing one (return 409 response)
 
+*Decision:* We decided to go with #2 and create our own API. The new batch retrieval API endpoint will not be forward facing and be for **internal use only**. 
+
+### Design
+
+#### Secrets Provider Job flow
+
+As a small recap, the following explains the Secrets Provider Job flow
+
+1. The `cyberark-secrets-provider-for-k8s` runs as a Job and authenticates with the Conjur/DAP Follower using the Kubernetes Authenticator (`authn-k8s`).
+
+2. The `cyberark-secrets-provider-for-k8s` reads all Kubernetes secrets required by the applications in the same namespace.
+
+3. For each mapped Kubernetes secret, 
+
+   a. The `cyberark-secrets-provider-for-k8s` performs a Batch Retrieval request and retrieves Conjur/DAP secrets.
+
+   b. It updates the Kubernetes secret with the Conjur/DAP secret value.
+
+4. The `cyberark-secrets-provider-for-k8s` Job is completed.
+
+The new Batch Retrieval API will effect #2a. 
+
+### URI
+
+##### POST /secrets/batch
+
+This method retuns a list of secrets, their status code, and their value if an error did not occur. 
+
+```
+POST /secrets/batch
+
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+### Request Body
+
+The request body is a list of all variable IDs (paths to secrets in Conjur/DAP) that need to be retrieved. For example: 
+
+```
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+#### Headers
+
+```
+Content-Type: application/json
+```
+
+### Response
+
+| Code             | Description                                                  |
+| ---------------- | ------------------------------------------------------------ |
+| 200 OK           | All secret values were retrieved successfully                |
+| 400 Bad Request  | Server cannot process request due to client error (invalid input parameters or logical validations) |
+| 401 Unauthorized | The request lacks valid authentication credentials           |
+| 403 Forbidden    | The authenticated user lacks the necessary privilege         |
+| 404 Not Found    | At least one variable does not exist, or at least one variable does not have any secret values |
```

Comment:
in this case will the secrets that were found be returned in the response body?

if so - isn't it wrong to return 404 but still return data?
if not - isn't this exactly what you wanted to fix?

Reply:
> in this case will the secrets that were found be returned in the response body?

Yes

> if so - isn't it wrong to return 404 but still return data?
Data meaning the error message if a 404 response was returned. Upon success, the value of the secret will be returned. Upon failure, the error message (Not Found) will be substituted for the the value, not the data of the secret.
Why is this wrong? Do you expect value to be `""` instead of `Not Found`?

Reply:
just to be clear. a user may get a response with code 404 and the following body:
```
{
  "successSecret1":[
     {
       "status": 200
       "value": "superSecret"
     }
  ]
  "failSecret1":[
    {
      "status": 404
      "value": "NotFound"     
    }
  ]
}
```

it seems weird to me that i get a response body with secrets where the error code is 404.

Reply:
but this is also a reasonable response:
```
HTTP/1.1 200 OK
{
  "successSecret1":[
     {
       "status": 200
       "value": "superSecret"
     }
  ]
  "failSecret1":[
    {
      "status": 403
      "value": "Forbidden"     
    }
  ]
}
```

so 200 in case one is forbidden but 404 in case one is not found.

Reply:
See here: https://docs.microsoft.com/en-us/graph/json-batching#:~:text=The%20status%20code%20on%20a,requests%20inside%20the%20batch%20succeeded

Regardless of the outcome of the secrets that are being fetched, if we can parse the body of the response we return a 200 status code because the body was parsable and "a 200 status code on the batch response does not indicate that the individual requests inside the batch succeeded."

Reply:
so this goes for 403 as well or only for 404? the behaviour is inconsistent between the 2.

Reply:
As we spoke on the phone, there are two different responses for this new batch endpoint. The overall response and the internal sub-responses